Add fast-path parameter binding

jakelishman opened the pull request:
The new `ParameterExpresion.bind_all` is a fast path for producing a numeric result.  This has a couple of advantages over `ParameterExpression.bind`:

- we can allocate far fewer Python objects, since we do not need to allocate new versions of all the additional data that goes along with a complete `ParameterExpression` object; we know the output must be numeric.

- we have no historical API reasons to scan through the incoming mapping to search for invalid keys or values; this is a huge performance improvement for the case of using the same mapping to bind many different expressions (though there are ways that `ParameterExpression.bind` could be improved and its interface safely evolved to remove the worst cases of the scaling).

There is still a sizeable amount of performance gain to be had, because the interfaces in the Rust-space `ParameterExpression` and `SymbolExpr` require rather more heap allocations per binding call than are entirely necessary.  That said, this already provides a large performance improvement, and a complexity improvement for the case of a large values dictionary being used many times.

<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary



### Details and comments

Fix #14471.

Currently in draft because there's no tests - I'm just putting it up so Sam and Ian from #14471 can test it out for their use case.  For the explicit example in that issue, a complete comparison on my machine:

<details><summary>Out of date timings</summary>
```python
In [1]: from qiskit.circuit import Parameter, ParameterExpression
   ...:
   ...: N: int = 100_000
   ...:
   ...: parameter_values = {Parameter(f"th_{i}"): 1 for i in range(N)}
   ...: parameter_values[param := Parameter("my_param")] = 1
   ...:
   ...: print("Using the specialised `Parameter` methods:")
   ...: %timeit param.bind(parameter_values, allow_unknown_parameters=True)
   ...: %timeit param.bind_all(parameter_values)
   ...:
   ...: print("Using the general `ParameterExpression` methods:")
   ...: %timeit ParameterExpression.bind(param, parameter_values, allow_unknown_parameters=True)
   ...: %timeit ParameterExpression.bind_all(param, parameter_values)
```
```text
Using the specialised `Parameter` methods:
32.8 ms ± 535 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)
68.5 ns ± 0.767 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
Using the general `ParameterExpression` methods:
34.2 ms ± 1.29 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
974 ns ± 8.14 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```
</details>

Cryoris reviewed (commented):
Would it make sense to expose a similar `bind_all` to the circuit, where we know that all parameters are being bound and we can call this fast path?

And another question actually: what do you think about changing Qiskit's default to _not check_ for parameter existence? Indeed we could be faster and maybe we ought to speed up the default behavior. I think most symbolic engines do not have these existence checks we added.

jakelishman commented:
`QuantumCircuit.assign_parameters` uses almost completely separate mechanisms to fast-path itself, and will have no need to interact with the Python-space methods of `ParameterExpression` once you've finished moving it to Rust properly.  This is addressing a particular use in the Primitives (think: parameter binding in a separate `EstimatorPub`).  They already set `allow_unused_parameters=True` in the calls to `bind`, but there's tonnes of other inefficiencies in there, not to mention that `ParameterExpression.bind` is API-stability required to return a `ParameterExpression`, even if it's fully bound.

There's a bunch of ways to reduce the overhead and the unnecessary looping through the parameter dictionaries in `ParameterExpression.bind`, but even if we did that, it'd still not be as fast as this method, so it can be in follow-ups if it's needed.

jakelishman committed:
Add fast-path parameter binding

The new `ParameterExpresion.bind_all` is a fast path for producing a
numeric result.  This has a couple of advantages over
`ParameterExpression.bind`:

- we can allocate fewer objects, since we do not need to allocate output
  name names, or other deeper metadata involved in the
  `ParameterExpression` struct

- we have no historical API reasons to scan through the incoming mapping
  looking for invalid keys or values.  This is a huge performance
  improvement for the case of using the same mapping to bind many
  different expressions (though there are always ways that
  `ParameterExpression.bind` could be improved in-place without breaking
  its API).

There is still a sizeable amount of performance gain to be had, because
the interfaces in the Rust-space `ParameterExpression` and `SymbolExpr`
require rather more heap allocations per binding call than are entirely
necessary.  That said, this already provides a large performance
improvement, and a complexity improvement for the case of a large values
dictionary being used many times.

jakelishman commented:
I rewrote this against the new Rust-space parameters.  The timings of the new methods got worse, though they're still _hugely_ better than the state of affairs without this PR - they move the timings from being on the order of 10ms-30ms per call to being on the order of 1-2µs, so 10,000x speedup for the particular problem size here:
```python
from qiskit.circuit import Parameter, ParameterExpression

N: int = 100_000

parameter_values = {Parameter(f"th_{i}"): 1 for i in range(N)}
parameter_values[param := Parameter("my_param")] = 1

print("Using the specialised `Parameter` methods:")
%timeit param.bind(parameter_values, allow_unknown_parameters=True)
%timeit param.bind_all(parameter_values)

print("Using the general `ParameterExpression` methods:")
%timeit ParameterExpression.bind(param, parameter_values, allow_unknown_parameters=True)
%timeit ParameterExpression.bind_all(param, parameter_values)
```
```text
Using the specialised `Parameter` methods:
13 ms ± 894 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
587 ns ± 3.9 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
Using the general `ParameterExpression` methods:
33.7 ms ± 842 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)
1.84 μs ± 38.2 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

I imagine there's two parts to the slow-down compared to the previous Python-based implementation:
- general PyO3 overhead on some very cheap function calls (but this isn't much - just on the super fast stuff)
- the new Rust-space `ParameterExpression` and `SymbolExpr` need a surprising amount of internal allocations to go back and forth between themselves and Python space, and to call methods like (the internal) `bind`.  I think the path I'd hit through the previous Python-space implementation didn't involve as much overhead here (though it did still involve a non-trivial amount).

jakelishman committed:
Remove unnecessary lifetime

kevinhartman reviewed (commented):
This looks good to me.

I know that the Rust `ParameterExpression::bind` method also does some error checking on the result, e.g. testing for infinity and NaN. Is that something we should do here too?

jakelishman commented:
Personally I think those existing error checks on finiteness of the output are mostly just a time-sink - infinity is a very valid (and necessary) floating-point value, while `NaN` is already a sufficiently visible poisoning of the data (c.f. an exception) and iirc we arbitrarily claim that `NaN` means `ZeroDivisionError`?  This is supposed to be a very fast-path method, so my biggest concern is its speed, but if you think the consistency is more important, it's not a huge sacrifice to make it work the same, now we're in compiled code.

kevinhartman reviewed (commented):
I think it's fine without having the same behavior. For clarity it might be helpful to add a blurb to the `bind_all` docstring to say that "unlike `bind`, `NaN` and inf are in the range of expected outputs for this method".

jakelishman committed:
Add comment on non-finite values

jakelishman commented:
Done in a5e83ce18f.

kevinhartman reviewed (approved):
LGTM, thanks!
